aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -221,6 +243,48 @@ module aptos_framework::delegation_pool {
         distribute_commission_events: EventHandle<DistributeCommissionEvent>,
     }
 
+    struct VotingRecordKey has copy, drop, store {
+        voter: address,
+        proposal_id: u64,
+    }
+
+    struct VotingPowerDelegation has copy, drop, store {
+        // The current voter address of this delegator.
+        voter: address,
+        // The change of voter address won't take effects until the current lockup period ends. Record the new voter
+        // address and update `voter` to `voter_next_lockup` when that lockup period has already ended.
+        voter_next_lockup: address,
+        // locked_until_secs of the stake pool when last time this struct is updated.
+        // If <locked_until_secs of the stake pool> != last_locked_until_secs, it means that a lockup period has passed.
+        // This is slightly different from ObservedLockupCycle because ObservedLockupCycle cannot detect if a lockup
+        // period is passed when there is no unlocking during the lockup period .
+        last_locked_until_secs: u64,
```

Comment:
This would break though if we ever introduce the ability for delegation pools to reset their lockup as although the lockup period might not have passed, the delegation pool's new locked until secs would be higher than what's stored here. Were there any other alternatives that you considered here?

Reply:
If we support resetting the lockup period , the time of voter change needs to be postponed to the new `locked_until_secs`. Otherwise let's consider a case: if A delegates to X then changes it to Y, X is A's voter before before the old `locked_until_secs` then Y is A's voter after the old `locked_until_secs`. X could vote on proposal 1 before the old `locked_until_secs` then Y could vote on proposal 1 after the old `locked_until_secs`. A's voting power will be used twice.
To implement this, we will need to record the start and the end of a lockup period. Let's say the start/end is a/b at the moment. If `last_locked_until_secs` is within [a, b], we can know the lockup period for that record hasn't ended yet. 

To maintain [a, b], we need:
- If the current lockup period ends normally and the new `locked_until_secs` is c. The start/end becomes [b, c].
- If the delegation pool resets its lockup period and the new `locked_until_secs` is c. The start/end becomes [a, c].

The solution could be plugged into the module when we want to support resetting. Do you think if we should add [a, b] now?

Reply:
That sounds rather complex. Is there no other alternative?

Reply:
As I understand, delegated voting power is delayed for a new voter to the next lockup cycle in order to ensure the proposal expires and double voting cannot happen if original voter already voted.
If lockup time can be refreshed on StakePool then each DelegatedVotes object should somehow detect that and update its `last_locked_until_secs`
However, the observed lockup cycle can be forced to be in sync with the real lockup cycle on StakePool. At `delegate_voting_power`:
1. Sync pool
2. Redeem shares worth 1 octa from caller if the resource account doesn't have any shares in the pending inactive pool
3. Buy shares for resource account for the 1 extracted octa, it will get 1 shares back (10 ^ (scaling factor set on pool_u64) shares really)
4 Resource account cannot reactivate this stake
5 Lockup ends on Stakepool, it inactivates 1 octa which is detected at next synchronization on delegation pool
This ensures OLC advances each lockup when someone changed their delegated voter and this change is guaranteed to be applied. So `last_locked_until_secs` can be replaced by a OLC index instead and not be affected by lockup expiration refreshing.
There will always be one shares pool created to host only 1 octa as buying pending-inactive shares for the resource account triggers the withdrawal of its inactive stake which empties the former shares pool which, in turn, gets deleted.

Reply:
@alexfilip2  Hmm I'm not sure I understand. Few questions:
1. What is octa?
2. What does "resource account" refer? NULL_SHAREHOLDER or the stake pool resource accounts?

Reply:
1. One unit of APT token - 10^-8
2. The resource account storing the DelegationPool, GovernanceRecords etc. It could be the NULL_SHAREHOLDER too, basically any account that cannot be accessed directly and reactivate stake

Reply:
Think more about it. 1 octa from caller might be problematic. There could be some edge cases:
1. the caller doesn't have any stake.
2. 1 octa might make the caller below `MIN_COINS_ON_SHARES_POOL`.

Reply:
after some more thought, I think that 1 octa doesn't have to come from the caller. Step 2 can be replaced by stake::unlock(1) - you get 1 coin from the active shares pool and then move to step 3 to buy pending-inactive shares for the resource account for 1 octa.

Reply:
Or: if a delegator wants to change its delegated voter, they should unlock some stake (the minimum possible one). If they reactivate their stake before the lockup expires then OLC won't advance and the new voter remains pending for another lockup cycle at least. This is less UX efficient, but it requires no complex logic for migrating to OLC indices instead of lockup timestamps

Reply:
@alexfilip2 The accounting model has already been very complicated so I prefer not adding more complexity. As long as we change behavior of the staking part, it brings lots of implication and edge cases. e.g. all delegators exit from a delegation pool/ maintaining `MIN_COINS_ON_SHARES_POOL`/ how flags can gate the logic. I think we should keep the governance part as standalone as possible.